src/rules/noTrailingWhitespaceRule.ts

```diff
@@ -72,53 +73,49 @@ export class Rule extends Lint.Rules.AbstractRule {
 }
 
 function walk(ctx: Lint.WalkContext<IgnoreOption>) {
-    let lastSeenWasWhitespace = false;
-    let lastSeenWhitespacePosition = 0;
-    utils.forEachTokenWithTrivia(ctx.sourceFile, (fullText, kind, range) => {
-        if (kind === ts.SyntaxKind.NewLineTrivia || kind === ts.SyntaxKind.EndOfFileToken) {
-            if (lastSeenWasWhitespace) {
-                reportFailure(ctx, lastSeenWhitespacePosition, range.pos);
+    const possibleFailures: ts.TextRange[] = [];
+    const sourceFile = ctx.sourceFile;
+    const text = sourceFile.text;
+    for (const range of getLineRanges(sourceFile)) {
```

Comment:
It can be done without `indexOf` because enough information is returned by `exec`. Then it's easier to use `len` instead of `end`, so you end up with:

```ts
function walk(ctx: Lint.WalkContext<IgnoreOption>) {
    const { options, sourceFile } = ctx;
    const possibleFailures = getPossibleFailures(sourceFile.text);
    if (possibleFailures.length === 0) {
        return;
    }

    const excludedRanges = options === IgnoreOption.None
                           ? getTemplateRanges(sourceFile)
                           : getExcludedRanges(sourceFile, options);
    for (const { pos, len } of possibleFailures) {
        if (!excludedRanges.some((range) => range.pos < pos && pos < range.end)) {
            ctx.addFailureAt(pos, len, Rule.FAILURE_STRING, ctx.createFix(Lint.Replacement.deleteText(pos, len)));
        }
    }
}

function getPossibleFailures(text: string): Array<{ pos: number, len: number }> {
    const result = [];
    const rgx = /[^\S\n]+(\r?\n|$)/g;
    let match: RegExpExecArray | null;
    while (match = rgx.exec(text)) {
        const matched = match[0];
        result.push({ pos: match.index, len: matched.endsWith("\n") ? matched.length - 1 : matched.length });
    }
    return result;
}
```